### PR TITLE
GP-4 ER

### PR DESCRIPTION
**Pull request by besanalqadi**
This is the ER diagram. Please review it so i can implement it into schema

### CODE REVIEW

File: ER.png
Comment (ahmad-abdelraheem) [inline ER.png:1]: emit the name before 'id' such in User-id, Pos-id. it's looks like ```book.book``` !

**Review by ahmad-abdelraheem — COMMENTED** [review]
Great work, @besanalqadi! I have a few comments.

## Naming

- Omit the entity name before 'id' (e. g. User_id -> Id).
- Let's try to avoid ambiguous names such as has a. I recommend these names for the relationships:
    - ``` [User - Position]``` : Aim to, Target.
    - ``` [User - Roadmap]``` : Follow.
    - ``` [Skill - Position]``` : Cover.
    - ``` [Skill - Roadmap]``` : Includes.
    - ``` [Position - Roadmap]``` : Achieved by.
    - ``` [Position - Domain]``` : Relates to.

> Naming is very important to help the team understand the ERD and our business overall.

## Relationship

- The ``` [Roadmap - Domain] ``` relation is not a direct one, so no need to chart it. A domain includes positions, and a position has a roadmap to achieve.
- There is a missing relationship: ``` [Roadmap - Position] ```. A position has one/many roadmaps to achieve.
- The ``` [Roadmap - Skill] ``` relationship has ```order: number``` and ```isDynamicOrder: boolean``` attributes to define the order of skills in the roadmap and whether the student can temporarily skip (reorder) it.

> The ``` [Skill - Position] ``` relationship should be ```M-N```. In this case, we have to map it to a separate table. This will make the DB more complicated, and I believe it's not worth it because there is no need to link  them. Even if we want to, we can easily query it using an ``` SQL join``` . I am voting to omit this relationship, but you can confirm that with Mr. Ayman Dmour.

## Cardinality

- ``` [User - Skill] ``` -> N-M: A user has many skills. A skill is possessed by many users.
- ``` [User - Position] ``` -> N-1: A user aims for one position. A position is targeted by many users.
- ``` [User - Roadmap] ``` -> N-1: A user follows one roadmap at a time. A roadmap is followed by many users.
- ``` [Skill - Roadmap] ``` -> N-M: A skill is included in many roadmaps. A roadmap includes many skills.
- ``` [Skill - Position] ``` -> N-M: A skill serves many positions. A position is served by many skills.

    > This relationship is open to removal as discussed earlier.

- ``` [Roadmap - Position] ``` -> N-1: A roadmap helps to achieve one position. A position is achieved by many roadmaps.
- ``` [Position - Domain] ``` -> N-M: A position relates to many domains. A domain includes many positions.

    > I know it's an uncommon case but may still occur, such as a DevOps position included in multiple IT domains.

I hope I haven't forgotten anything. Please let me know if you have any questions. Great job again!

**Review by TuqaSa — COMMENTED** [review]
Building a branching roadmap database involves structuring your data so that users can follow different paths depending on the choices they make at certain stages of the roadmap. In this case, each roadmap might have multiple paths, and the user can decide to follow a specific branch based on preferences or skillset.

Here’s how to structure the database for a branching roadmap:

### Key Concepts:
- **Roadmap**: Represents the overall journey to achieve a position.
- **Stage**: Represents individual steps or milestones in the roadmap.
- **Branch**: A point where the roadmap splits into different paths.
- **Skill**: Represents the knowledge or ability required at a specific stage.
- **Position**: Represents the end goal of the roadmap.
  
### Database Structure:

#### Entities:
1. **Roadmap**: This will be the starting point of the journey.
   - Attributes: `Id`, `Name`, `Description`, `Position_Id`
  
2. **Stage**: Represents each stage in the roadmap.
   - Attributes: `Id`, `Name`, `Description`, `Order`, `Roadmap_Id`
   - Relationship: Each roadmap has many stages.
  
3. **Branch**: Represents decision points that lead to different paths in the roadmap.
   - Attributes: `Id`, `Stage_Id`, `Next_Stage_Id`, `Condition`
   - Relationship: Each stage can have multiple branches that lead to different next stages.

4. **Skill**: Skills required at certain stages.
   - Attributes: `Id`, `Name`, `Description`
   - Relationship: Many-to-many with `Stage` (a stage requires many skills, and a skill can be part of many stages).

5. **Position**: The final career goal.
   - Attributes: `Id`, `Name`, `Description`
   - Relationship: A roadmap is linked to one position.

#### Relationships:
1. **Roadmap - Position**: One-to-many (A roadmap leads to one position, but a position can have multiple roadmaps).
2. **Roadmap - Stage**: One-to-many (A roadmap has many stages).
3. **Stage - Skill**: Many-to-many (A stage requires many skills, and a skill can be required at many stages).
4. **Stage - Branch**: One-to-many (A stage can branch into multiple next stages).
5. **Branch - Stage**: One-to-one (A branch leads to the next stage based on conditions).

#### Example Database Tables:

1. **Roadmap**: 
   - `Id`, `Name`, `Description`, `Position_Id`
   
2. **Stage**:
   - `Id`, `Name`, `Description`, `Order`, `Roadmap_Id`
   
3. **Branch**:
   - `Id`, `Stage_Id`, `Next_Stage_Id`, `Condition`
   
4. **Skill**:
   - `Id`, `Name`, `Description`
   
5. **Position**:
   - `Id`, `Name`, `Description`

6. **Stage_Skill** (Join table for the many-to-many relationship between `Stage` and `Skill`):
   - `Stage_Id`, `Skill_Id`

### Workflow:
1. A **User** selects a **Roadmap** to follow.
2. The roadmap starts with **Stage 1** and progresses to the next stage.
3. At specific stages, the roadmap may have a **Branch** that leads to multiple different **Next Stages**.
4. The user chooses a branch based on the **Condition** (e.g., whether they have a certain skill or completed a prerequisite).
5. The **Skills** associated with each stage must be acquired or demonstrated before proceeding.

This branching structure allows the user to follow different paths flexibly. The **Branch** entity, with its `Condition`, enables dynamic decisions at key points in the roadmap.

**Comment by ahmad-abdelraheem** [comment]
> Building a branching roadmap database involves structuring data so that users can follow different paths depending on the choices they make at certain stages of the roadmap. Each roadmap may have multiple paths, allowing users to follow a specific branch based on preferences or skillsets.

In follow-up to @TuqaSa’s review, the idea of having a "Stage" entity is good, but we encounter the same problem regarding the ordering and mapping of stages after the first branch. To clarify the issue, I'll provide this example to guide our discussion toward a more effective and straightforward path:

image

In our prototype, we will save this roadmap as follows:

### Position
| Id  | Title    | Description |
|-----|----------|-------------|
| 1   | Front-End | Text        |

### Roadmap
| Id  | Title    | Description      |
|-----|----------|------------------|
| 1   | Angular  | Angular roadmap  |
| 2   | React    | React roadmap    |

### Skills
| Id  | Title     | Description                                                                          |
|-----|-----------|--------------------------------------------------------------------------------------|
| 1   | HTML      | HyperText Markup Language                                                            |
| 2   | CSS       | Cascading Style Sheets                                                               |
| 3   | JavaScript| Scripting language that executes in the browser                                       |
| 4   | TypeScript| A superset of JS with additional features like decorators, strong typing, and compile-time error detection |
| 5   | React     | A widely used JS library that leverages JSX and functions to build components         |
| 6   | Angular   | A popular TypeScript framework used to build components                              |
| 7   | NextJS    | Full-stack framework built on React                                                   |
| 8   | RemixJS   | Placeholder, as I’m unsure of the exact details at the moment                         |

### Roadmap - Skills
> This table should map the relationship between the Roadmap and Skills, with an `order` attribute.

| Roadmap Id | Skill Id | Order |
|------------|----------|-------|
| ngRoadmap  | HTML     | 1     |
| actRoadmap | HTML     | 1     |
| ngRoadmap  | CSS      | 2     |
| actRoadmap | CSS      | 2     |
| actRoadmap | JavaScript| 3    |
| ngRoadmap  | TypeScript| 3    |
| actRoadmap | React    | 4     |
| ngRoadmap  | Angular  | 4     |
| actRoadmap | NextJS   | 5     |
| actRoadmap | RemixJS  | 5     |

> Note: Skill Id is replaced with the skill name for clarity. Roadmap Id is replaced with `ngRoadmap` for Angular and `actRoadmap` for React.

This is the scenario for how the roadmap will be rendered on the client-side:

1. The first and second skills in both front-end roadmaps are identical, so we render one node with no branches for each skill.
2. The third skill differs between the two roadmaps, so we render two nodes branching from the second level (CSS branches into JS and TS).
   > Up to this point, there’s no issue.
3. The skills at order 4 also differ, but there’s only one skill for each roadmap at this level, so we continue along the branches created at level 3. We map the skills at level 4 to the skills at level 3 by tracking the `Roadmap ID`.

Now, here’s the problem: Imagine we omit the roadmap relation (table). How could we map the fourth level to the third? For example, how do we know where to place the "Angular" skill? Should it follow "JavaScript" or "TypeScript"? Without the `Roadmap ID`, there’s no way to tell. We could introduce another key like `previous skill` and `next skill` to implement something akin to a linked list, but is that really better than using the roadmap relation?

I’m pleased with this constructive discussion and look forward to reaching an even better outcome by the end of it.

> Note that there are a more complex cases for the roadmap that we can discuss later on. this is just an example. I will put some examples below.

image

> In the example above, we will not re merge the paths because ``` VSC (git) ``` node Occur in different order on each roadmap. so the result will be like below.

image

___

This issue is cumulative and will continue to build up along the entire path as the following.

### What we should Render : 

image

### What we will actually render : 

image

> Anyway, this issue will not flow to the top (will not be displayed for the end-user since we will omit any thing related to path 2 (roadmap 2) when user select path 1. but still it's saved in this way and will displayed in this way before selecting a path.

@besanalqadi @TuqaSa @mahmoud661 @kafkacs

**Comment by kafkacs** [comment]
based on the schema which ahmad gave and the disction we had in the meeting, maybe we can add a steps table that includes the current step with its id and the next step and have the next step can route to a new road map for example:



road map schema

**this is the first table where we store the user skills needed to the road map and its default starting step in the steps table which I will provide below**


steps schema

**each step in the roadmap covers a different area, such as HTML & CSS, JavaScript, or frameworks like React. Some steps offer choices like (react, angular... ect). the next step refers to a step in the the same table which that step can show its next stage and so on.**

**and on each step we can add a suggestion of course to learn**


I apologize for the misspelling of some words and the unorganized comment its my first one.
thank you.
@ahmad-abdelraheem @TuqaSa @mahmoud661 @besanalqadi

**Comment by mahmoud661** [comment]
After conducting research, I discovered a package for React called **React Flow**. This library is designed for building node-based applications and allows for easy visualization of data flows, making it an excellent choice for representing roadmaps.

## Purpose of Using React Flow
React Flow helps create interactive flow diagrams by allowing users to connect nodes (representing stages) visually. This visualization aids in understanding the relationships and progressions within a roadmap .

## Data Structure
To effectively utilize React Flow , I organized the data as follows:

### 1. Roadmaps Table

| Column Name  | Data Type   | Description                                           |
|--------------|-------------|-------------------------------------------------------|
| `id`         | SERIAL      | Unique identifier for the roadmap (Primary Key)      |
| `name`       | VARCHAR(255)| Name of the roadmap                                   |
| `description`| TEXT        | Description of what the roadmap covers                |


### 2. Branches Table

| Column Name  | Data Type   | Description                                           |
|--------------|-------------|-------------------------------------------------------|
| `id`         | SERIAL      | Unique identifier for the branch (Primary Key)       |
| `roadmap_id` | INT         | Foreign key referencing the Roadmaps table            |
| `name`       | VARCHAR(255)| Name of the branch                                    |
| `description`| TEXT        | Description of the branch                             |
| `source`     | INT         | The ID of the source stage (foreign key referencing the Stages table) |
| `target`     | INT         | The ID of the target stage (foreign key referencing the Stages table) |


### 3. Stages (Nodes) Table

| Column Name  | Data Type   | Description                                           |
|--------------|-------------|-------------------------------------------------------|
| `id`         | SERIAL      | Unique identifier for the stage (Primary Key)        |
| `branch_id`  | INT         | Foreign key referencing the Branches table            |
| `name`       | VARCHAR(255)| Name of the stage                                     |
| `description`| TEXT        | Description of what the stage entails                 |
| `position`   | JSON        | JSON object to store the x and y coordinates          |


- **Roadmaps** can have multiple **Branches** (one-to-many relationship).
- Each **Branch** connects two **Stages** (source and target), which allows for defining the flow or progression between stages.
- **Stages** are still linked to their respective **Branches**.

after I test it i got somthing like this:

Recording 2024-10-25 154949 (1)


This how the branchs data will look like :
```
[
  { id: 'e1-2', source: '1', target: '2' },
  { id: 'e2-3', source: '2', target: '3' },
  { id: 'e3-4', source: '3', target: '4' },
  
  { id: 'e4-5', source: '4', target: '5'  }, 
  { id: 'e4-7', source: '4', target: '7'  }, 
  { id: 'e5-6', source: '5', target: '6' },
  { id: 'e7-8', source: '7', target: '8' },
];
```

and the nodes or stages :
```
[
  { id: '1', type: 'input', position: { x: 0, y: 0 }, data: { label: 'Learn HTML' } }, 
  { id: '2', position: { x: 0, y: 100 }, data: { label: 'Learn CSS' } }, 
  { id: '3', position: { x: 0, y: 200 }, data: { label: 'TypeScript Basics' } },
  { id: '4', position: { x: 0, y: 300 }, data: { label: 'Advanced TypeScript' } },
  
  // Framework choice nodes
  { id: '5', type: 'default', position: { x: -100, y: 400 }, data: { label: 'Learn JSX (React)' } },
  { id: '6', type: 'default', position: { x: -100, y: 500 }, data: { label: 'React Hooks' } },
  
  { id: '7', type: 'default', position: { x: 100, y: 400 }, data: { label: 'Angular Modules' } },
  { id: '8', type: 'default', position: { x: 100, y: 500 }, data: { label: 'Angular Services' } },
];
```

In this document, I have discussed how we will structure the roadmap. If there is anything I missed, please let me know.

**Comment by besanalqadi** [comment]
My solution is: 
-When there is a branching in the roadmap, ask the user about the next language or skill and store his answer in the skills table (the proposed skills are part of the roadmap displayed) The example shown shows the point:

Blank diagram

-because the skill  can has more than next position  so it will be a multivalued attribute ( the next position will become a table with a primary key Consists  the id of the position and the name of the next position together) and it have the skill id as a att  .

- the relationship between the skill table  and the next position is one to many relationship (the skill might has many next position ). 

-After storing the selected skill, we join the skill table with the next position table and see the next position for the selected skill that was previously stored in the skill table , the join will be in the next position. skill-id= skill-id , the  new table will show the name for the skill and its next position the roadmap will take the value of  the next position .

- making the roadmap Derived Attribute will help with Real-time accuracy .

**Comment by ahmad-abdelraheem** [comment]
In follow-up to @kafkacs and @mahmoud661 comments, I don't actually like the data redundant in these suggestions.

1. In @kafkacs's schema, we have grouped skills in the steps and any differ causes a new step. So will have a lot of duplicated skills in the steps table. 
2. In @mahmoud661's Schema, the problem actually in the ``` position ``` and in the skills store. We need to store the skills to measure user's progress in his selected roadmaps and other roadmaps if he want to shift his career or target position.

I will put this suggestion based on this discussion.

## Position

| Id | Name      |
| -- | --------- |
| 1  | Front-End |
| 2  | Back-End  |
| 3  | DevOps    |


## Roadmap

|  Id |       Name      | Position\_Id |
| :-: | :-------------: | :----------: |
|  1  | Angular Roadmap |       1      |
|  2  |  React Roadmap  |       1      |
|  3  |   Vue Roadmap   |       1      |
|  4  |   .Net Roadmap  |       2      |
|  5  | Node Js Roadmap |       2      |

## Major Skills

Includes the skill without details (without the steps to achieve the skill)
This table will help us to order the skill in the roadmap, track the keywords in job offers and track the progress in different roadmaps.

|  Id |    Name    |
| :-: | :--------: |
|  1  |    HTML    |
|  2  |     CSS    |
|  3  | JavaScript |
|  4  | TypeScript |
|  5  |   Angular  |
|  6  |    React   |
|  7  |   Vue.js   |
|  8  |     C#     |
|  9  |     OOP    |
|  10 |     Git    |

## Sub Skills

This table includes the sub skills or steps inside each major skill (such as the major concepts and skills to learn angular framework)
This table will help us display a guide for achieving a specific skills in the roadmap (detailed roadmap)

| Id | Skill_Id | Name                    | order |
| -- | --------- | ----------------------- | ----- |
| 1  | Angular   | Introduction to Angular | 1     |
| 2  | Angular   | Components              | 2     |
| 3  | Angular   | modules                 | 3     |
| 4  | Angular   | Directive               | 4     |

## Roadmap - Skills

This is the main table to fetch a roadmap, storing roadmaps in this way will help us if we want to display a specific roadmap (for users who's already decided their career (User need an angular roadmap not a front-end general roadmap).
So in selecting this option will allow us to display merged roadmap for a position and display a specific roadmap based on skills which is the most common roadmaps. 

| Roadmap Id      | Skill Id   | Order | is it able to skip? | is it required? |
| --------------- | ---------- | ----- | ------------------- | --------------- |
| angular roadmap | HTML       | 1     | false               | true            |
| angular roadmap | CSS        | 2     | false               | true            |
| angular roadmap | JavaScript | 3     | false               | true            |
| angular roadmap | TypeScript | 4     | false               | true            |
| angular roadmap | Angular    | 5     | false               | true            |
| react roadmap   | HTML       | 1     | false               | true            |
| react roadmap   | CSS        | 2     | false               | true            |
| react roadmap   | JavaScript | 3     | false               | true            |
| react roadmap   | TypeScript | 4     | true                | false           |
| react roadmap   | React      | 5     | false               | true            |

After processing the date that we will for these schema, I have produced this roadmap
```
[
  [{ skill: "HTML", previousSkill: null, nextSkill: ["CSS", "CSS"] }],
  [
    {
      skill: "CSS",
      previousSkill: "HTML",
      nextSkill: ["JavaScript", "JavaScript"],
    },
  ],
  [
    {
      skill: "JavaScript",
      previousSkill: "CSS",
      nextSkill: ["TypeScript", "TypeScript"],
    },
  ],
  [
    {
      skill: "TypeScript",
      previousSkill: "JavaScript",
      nextSkill: ["Angular", "React"],
    },
  ],
  [
    ({ skill: "Angular", previousSkill: "TypeScript", nextSkill: null },
    { skill: "React", previousSkill: "TypeScript", nextSkill: null }),
  ],
]
```

@besanalqadi @kafkacs @mahmoud661 @TuqaSa

**Comment by ahmad-abdelraheem** [comment]
I’m considering a different approach to storing the roadmaps. Here’s an example structure for front-end roadmaps:

1. **Domain Roadmap**: Groups foundational skills common to all positions in the domain. For the web domain, this would include skills like Internet fundamentals, Git, GitHub, etc.
  
2. **Position Roadmap**: Groups all basic skills required across specific positions within the domain. For the front-end position, this would cover HTML, CSS, JavaScript, TypeScript, Bootstrap, etc.

3. **Advanced Roadmap**: A detailed, advanced roadmap that varies significantly between different frameworks (e.g., React or Angular).

With this structure, a user would complete the basics of their selected domain first, allowing flexibility to switch between positions within the same domain. After completing the basics, they would proceed to the skills relevant to their chosen position. Once the basics are complete, we would display advanced roadmap options for the user to choose from.

This approach makes it easier for users to choose between frameworks like Angular and React, rather than selecting a skill without understanding its impact on their overall path.

@besanalqadi @kafkacs @mahmoud661 @TuqaSa

**Comment by mahmoud661** [comment]
I think we’re overcomplicating this feature. Our goal is to provide a straightforward roadmap for the student.we keep mentioning roadmaps for frameworks in web development, but we need to remember that there are other paths beyond just web development, and not everyone will be using frameworks. If we allow skills to appear in multiple roadmaps, we can track what a student has completed by adding those skills to their profile. This way, even if a skill is duplicated in different roadmaps, we’ll have a clear record of what the student has finished. Let’s keep it simple , so we can implement the roadmap generator.

**Comment by ahmad-abdelraheem** [comment]
> I think we’re overcomplicating this feature. Our goal is to provide a straightforward roadmap for the student. We keep mentioning roadmaps for frameworks in web development, but we need to remember that there are other paths beyond just web development, and not everyone will be using frameworks. If we allow skills to appear in multiple roadmaps, we can track what a student has completed by adding those skills to their profile. This way, even if a skill is duplicated in different roadmaps, we’ll have a clear record of what the student has finished. Let’s keep it simple, so we can implement the roadmap generator.

I completely agree with @mahmoud661 . Let's keep the ER as it is, just adding a sub-skill to provide more details for the roadmap when a user requests it.

@besanalqadi , can you please apply these changes and request a review from the entire team? We will consider the approvals for this PR as a vote. Please provide a time estimation for that.

**Comment by besanalqadi** [comment]
The new change will be on the ER that was attached for the first time, right?
@ahmad-abdelraheem

**Comment by besanalqadi** [comment]
Blank board (2)
I hope from everyone to review the er. 
(A simple note the name plus id -skill are the key to the sub-skill)
@ahmad-abdelraheem @mahmoud661 @TuqaSa @kafkacs

**Comment by besanalqadi** [comment]
[gc.pdf](https://github.com/user-attachments/files/17549691/gc.pdf)
I hope from everyone to review the Gantt chart.

**Comment by besanalqadi** [comment]
[nd.pdf](https://github.com/user-attachments/files/17549711/nd.pdf)
I hope from everyone to review the network diagram.
@ahmad-abdelraheem @mahmoud661 @TuqaSa @kafkacs

**Comment by besanalqadi** [comment]
[project-doc.docx](https://github.com/user-attachments/files/17550066/project-doc.docx)
This is the first chapter of the project's documentation.
@ahmad-abdelraheem @mahmoud661 @TuqaSa @kafkacs

**Comment by kafkacs** [comment]
in the Gantt chart that @besanalqadi provided shouldn't we discus about adding a QA in the chart plus add a section for the User Acceptance Testing? and I think we should add more time for the design phase and lower the Analysis since design would probably take the most time out of all? 
however great work besan! thank you

**Comment by ahmad-abdelraheem** [comment]
> [nd.pdf](https://github.com/user-attachments/files/17549711/nd.pdf) I hope from everyone to review the network diagram. @ahmad-abdelraheem @mahmoud661 @TuqaSa @kafkacs


@besanalqadi  please make a ticket and PR for the gannet chart, network diagram and first chapter to review. lets make this repo as clean as possible.

**Comment by ahmad-abdelraheem** [comment]
> The new change will be on the ER that was attached for the first time, right? Can you add more details about the sub-skill? Is it an Entity? What is its relationship with the rest of the system Entities? @ahmad-abdelraheem

Sure, the Sub skill is a weak entity that describe the steps and concepts inside each skills. If we consider that JavaScript is a skill, the sub skills will be :
- JavaScript Introduction
- Variables.
- if, else statement
- switch statement
- for loop
- while loop
- fetching loops
- array
- string
and so on.

this entity will help us to display more details about how to cover each skill in the roadmap (detailed roadmap).

please let me know if you need any more details. please push a commit with the changes and request review again to close this PR.

appreciate your effort @besanalqadi .

**Comment by ahmad-abdelraheem** [comment]
@besanalqadi , Can we use another notation please ?

image

> Zero mean ``` Partial ``` such as the dotted line in your notation.

___

image

> This entity style is clearer and provides more details, and it's also more familiar, as this is how most DBMS generate ERDs.

**Review by ahmad-abdelraheem — APPROVED** [review]

**Review by besanalqadi — COMMENTED** [review]
done

**Review by TuqaSa — APPROVED** [review]
This is more clear and put together

**Pull request merged**